### PR TITLE
Support core.abbrev 7-40

### PR DESCRIPTION
**Pull request by SocalNick**
### Your checklist for this pull request
🚨Please review the [guidelines for contributing](https://github.com/ruby-git/ruby-git/blob/master/CONTRIBUTING.md) to this repository.

- [x] Ensure all commits include DCO sign-off.
- [x] Ensure that your contributions pass unit testing.
- [x] Ensure that your contributions contain documentation if applicable.

### Description
We ran into an issue when we bumped to Ubuntu 18 (bionic) and started
seeing failures in process_full_diff because the index lines had 11
character SHA abbreviations. This should allow core.abbrev to be set to
anything between 7 (the default) and 40 (the max).

### CODE REVIEW

File: lib/git/diff.rb [hunk 1]
```diff
@@ -89,10 +89,10 @@ def binary?
       end
 
       def blob(type = :dst)
-        if type == :src
-          @base.object(@src) if @src != '0000000'
-        else
-          @base.object(@dst) if @dst != '0000000'
+        if type == :src && !/0{7,40}/.match(@src)
```
Comment (SocalNick) [inline lib/git/diff.rb]: Perhaps worth mentioning, I couldn't find a test that required the check for `@src != 0000000` to be there. Assuming there's some edge case it's needed for, I changed it to the regex version.

File: lib/git/diff.rb [hunk 2]
```diff
@@ -89,10 +89,10 @@ def binary?
...
+          @base.object(@src)
+        elsif !/0{7,40}/.match(@dst)
```
Comment (ngan) [inline lib/git/diff.rb]: Might be good to pull the regex out to a constant and use `match?` (instead of `match`) since you don't care to capture any values.

Comment (SocalNick) [reply lib/git/diff.rb]: Updated to use a constant for the regex. I initially tried `match?` but that isn't available in Ruby 2.3 which is in the test matrix for this project.

**Comment by ngan** [comment]
@jcouball any chance we can get this landed and released?

**Comment by tomgilligan** [comment]
i should have checked here first, ended up with my own version of a fix https://github.com/tomgilligan/ruby-git/commit/2d82185b482302efed0e19fd1b1aeee73d318715

worth mentioning that the minimum abbrev is apparently 4
https://git-scm.com/docs/git-config#Documentation/git-config.txt-coreabbrev

**Comment by tomgilligan** [comment]
also i think `NIL_BLOB_REGEXP` should be 'anchored'

File: lib/git/diff.rb [hunk 3]
```diff
@@ -72,6 +72,7 @@ def each(&block) # :yields: each Git::DiffFile in turn
     class DiffFile
       attr_accessor :patch, :path, :mode, :src, :dst, :type
       @base = nil
+      NIL_BLOB_REGEXP = /0{7,40}/.freeze
```
Comment (ngan) [inline lib/git/diff.rb]:
```suggestion
      NIL_BLOB_REGEXP = /\A0{7,40}\z/.freeze
```

File: lib/git/diff.rb (hunk 1 above)
Comment (tommy-gilligan) [reply lib/git/diff.rb]: seconded, there is no test for `@src` check

**Comment by SocalNick** [comment]
@tomgilligan thanks for the feedback! Ngan committed the change to anchor the `NIL_BLOB_REGEXP` and I added test to ensure this works with 4,7, and 40 (really anything between 4-40) for `core.abbrev`.

Not sure why the CI check isn't running, otherwise, is this good to merge?

**Comment by jcouball** [comment]
I re-ran the build and the status was updated.  Not sure why it wasn't updating before.

I have this one in my queue for review.  Sorry it has taken so long.

**Comment by ngan** [comment]
@jcouball bump

**Review by jcouball — APPROVED** [review]
Looks good.  Thank you for the thorough update!
